File: 11-性能分析/01-系统级剖析/OffWake调度延迟归因/offwake.go
// Off-Wake 调度延迟归因 —— Go 侧实现（无工具链，人工审查 + 机械核查）。
//
// 口径同 Python 版，来自 Brendan Gregg《Off-CPU Analysis》：
//   off-CPU time = 阻塞到重新开跑，含调度延迟（时间膨胀）；
//   唯一插桩点是 finish_task_switch() 结尾，在下一个线程的上下文中执行。
package main

import "fmt"

const (
	taskRunning        = 0
	taskInterruptible  = 1
	taskUninterruptible = 2
)

// Sample 是一段完整的 off-CPU 区间。
type Sample struct {
	TID        int
	Start, End float64
	State      int
	Stack      string
	WokenAt    float64 // <= 0 表示没有唤醒点
	WakerStack string
}

func (s Sample) OffCPU() float64 { return s.End - s.Start }

func (s Sample) Blocked() float64 {
	if s.WokenAt <= 0 {
		return s.OffCPU()
	}
	return s.WokenAt - s.Start
}

func (s Sample) SchedLatency() float64 {
	if s.WokenAt <= 0 {
		return 0
	}
	return s.End - s.WokenAt
}

func (s Sample) Inflation() float64 {
	b := s.Blocked()
	if b <= 0 {
		return 1
	}
	return s.OffCPU() / b
}

// Tracer 按原文伪代码实现 off-CPU 记账。
type Tracer struct {
	StateFilter int // <0 表示不过滤
	SleepTime   map[int]float64
	State       map[int]int
	WakeTS      map[int]float64
	WakerStack  map[int]string
	Samples     []Sample
}

func NewTracer(stateFilter int) *Tracer {
	return &Tracer{
		StateFilter: stateFilter,
		SleepTime:   map[int]float64{},
		State:       map[int]int{},
		WakeTS:      map[int]float64{},
		WakerStack:  map[int]string{},
	}
}

func (t *Tracer) OnWakeup(tid int, ts float64, wakerStack string) {
	t.WakeTS[tid] = ts
	t.WakerStack[tid] = wakerStack
}

// OnSwitchFinish 是 finish_task_switch(prev) 的结尾：给 prev 打起点、结算 cur。
func (t *Tracer) OnSwitchFinish(prevTID, prevState, curTID int, curStack string, ts float64) (Sample, bool) {
	t.SleepTime[prevTID] = ts
	t.State[prevTID] = prevState
	start := t.SleepTime[curTID]
	if start == 0 { // 0 表示没在睡 ⇒ 本次不是阻塞后的唤醒
		return Sample{}, false
	}
	state := t.State[curTID]
	s := Sample{
		TID: curTID, Start: start, End: ts, State: state, Stack: curStack,
		WokenAt: t.WakeTS[curTID], WakerStack: t.WakerStack[curTID],
	}
	delete(t.WakeTS, curTID)
	delete(t.WakerStack, curTID)
	t.SleepTime[curTID] = 0 // 原文：清零而不是删除
	t.Samples = append(t.Samples, s)
	return s, true
}

func (t *Tracer) Filtered() []Sample {
	if t.StateFilter < 0 {
		return t.Samples
	}
	var out []Sample
	for _, s := range t.Samples {
		if s.State == t.StateFilter {
			out = append(out, s)
		}
	}
	return out
}

// ByStack 按阻塞栈聚合 off-CPU 时间（折叠格式的 count 列）。
func (t *Tracer) ByStack() map[string]float64 {
	out := map[string]float64{}
	for _, s := range t.Filtered() {
		out[s.Stack] += s.OffCPU()
	}
	return out
}

// RequestSync 等价于 grep 请求上下文标记（线程池陷阱的解药）。
func (t *Tracer) RequestSync(marker string) []Sample {
	var out []Sample
	for _, s := range t.Filtered() {
		if contains(s.Stack, marker) {
			out = append(out, s)
		}
	}
	return out
}

// OffWakeKey 是 (唤醒者栈, 阻塞栈) 二元组。
type OffWakeKey struct{ Waker, Waiter string }

func (t *Tracer) OffWake() map[OffWakeKey]float64 {
	out := map[OffWakeKey]float64{}
	for _, s := range t.Filtered() {
		w := s.WakerStack
		if w == "" {
			w = "(unknown waker)"
		}
		out[OffWakeKey{w, s.Stack}] += s.OffCPU()
	}
	return out
}

func contains(s, sub string) bool {
	if sub == "" {
		return true
	}
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}

func main() {
	t := NewTracer(-1)
	t.SleepTime[1] = 1
	t.State[1] = taskUninterruptible
	t.OnWakeup(1, 11, "irq;net_rx_action")
	s, ok := t.OnSwitchFinish(3, taskRunning, 1, "app;read", 91)
	if ok {
		fmt.Printf("off_cpu=%.1f blocked=%.1f sched=%.1f inflation=%.2f\n",
			s.OffCPU(), s.Blocked(), s.SchedLatency(), s.Inflation())
	}

	t2 := NewTracer(taskUninterruptible)
	t2.SleepTime[1] = 1
	t2.SleepTime[2] = 1
	t2.SleepTime[3] = 1
	t2.State[1] = taskUninterruptible
	t2.State[2] = taskInterruptible
	t2.State[3] = taskRunning
	t2.OnSwitchFinish(9, taskRunning, 1, "io", 100)
	t2.OnSwitchFinish(9, taskRunning, 2, "sleep", 100)
	t2.OnSwitchFinish(9, taskRunning, 3, "spin", 100)
	fmt.Printf("全部 %d 条，--state 2 过滤后 %d 条\n", len(t2.Samples), len(t2.Filtered()))
	fmt.Printf("off-wake 键数=%d，按阻塞栈聚合只有 %d 个\n", len(t2.OffWake()), len(t2.ByStack()))
}

File: 11-性能分析/01-系统级剖析/OffWake调度延迟归因/offwake.py
"""Off-Wake 剖析：把 off-CPU 时间按「阻塞栈 + 唤醒者栈」联合归因 —— 模型层。

权威来源：Brendan Gregg, "Off-CPU Analysis"
https://www.brendangregg.com/offcpuanalysis.html

原文确立的事实（README §2 逐条对应）：
1. off-CPU time = 线程**阻塞**到**再次开始运行**之间的时间，**包含调度延迟**
   （"Off-CPU time consists of everything from when a thread blocked to when it
   began running again, including scheduler delay."）—— 这叫**时间膨胀**。
2. 插桩点是唯一的：上下文切换例程的结尾 `finish_task_switch(prev)`，
   在**下一个线程（next / cur）的上下文**中调用。原文伪代码：

       on context switch finish:
           sleeptime[prev_thread_id] = timestamp
           if !sleeptime[thread_id]
               return
           delta = timestamp - sleeptime[thread_id]
           totaltime[pid, execname, user stack, kernel stack] += delta
           sleeptime[thread_id] = 0

   之所以只测一次栈就够：**"Application stack traces don't change while off-CPU."**
3. 状态过滤：`TASK_RUNNING`(0) 是非自愿切换，`TASK_INTERRUPTIBLE`(1) /
   `TASK_UNINTERRUPTIBLE`(2) 才是通常关心的阻塞；`--state 2` 可只留 2。
4. 线程池陷阱：大部分阻塞时间落在「等待工作」的栈里（MySQL 的
   `do_nanosleep`/`futex_wait_queue_me`/`read_events`），要用请求上下文过滤
   （按折叠格式 `grep do_command`）只留请求同步路径。
5. 唤醒者视角：很多 off-CPU 栈只显示**阻塞路径**，真正的原因在**另一个线程**
   （发起 wakeup 的 waker）。工具 `wakeuptime` 测唤醒栈，`offwaketime` 把二者
   关联；宽度仍然是 off-CPU 时间（微秒）。
6. 火焰图左右顺序**没有意义**（"The left-to-right ordering has no meaning."）。
7. 开销：调度器事件可能每秒数百万，必须内核内聚合（eBPF），并警惕反馈回路；
   建议先只追踪 0.1 秒。
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Sequence, Tuple

TASK_RUNNING = 0
TASK_INTERRUPTIBLE = 1
TASK_UNINTERRUPTIBLE = 2

STATE_NAME = {0: "RUNNING", 1: "INTERRUPTIBLE", 2: "UNINTERRUPTIBLE"}


@dataclass
class Sample:
    """一段完整的 off-CPU 区间。"""

    tid: int
    start: float
    end: float
    state: int
    stack: str
    woken_at: Optional[float] = None
    waker_stack: str = ""

    @property
    def off_cpu(self) -> float:
        """阻塞到重新开跑的墙钟时间（原文口径，含调度延迟）。"""
        return self.end - self.start

    @property
    def blocked(self) -> float:
        """真正被阻塞的时间（不含运行队列等待）；无唤醒点时等于 off_cpu。"""
        if self.woken_at is None:
            return self.off_cpu
        return self.woken_at - self.start

    @property
    def sched_latency(self) -> float:
        """运行队列延迟 = 调度延迟，是时间膨胀的来源。"""
        if self.woken_at is None:
            return 0.0
        return self.end - self.woken_at

    @property
    def inflation(self) -> float:
        """膨胀比 = off_cpu / blocked，1.0 表示没有调度延迟。"""
        b = self.blocked
        return 1.0 if b <= 0 else self.off_cpu / b


class OffCpuTracer:
    """按原文伪代码实现的 off-CPU 记账器。"""

    def __init__(self, state_filter: Optional[int] = None):
        self.state_filter = state_filter  # None = 不过滤；2 = --state 2
        self.sleeptime: Dict[int, float] = {}
        self.state: Dict[int, int] = {}
        self.wakets: Dict[int, float] = {}
        self.waker_stack: Dict[int, str] = {}
        self.samples: List[Sample] = []

    # ---------------------------------------------------------------- 事件 ---
    def on_wakeup(self, tid: int, ts: float, waker_stack: str = "") -> None:
        """`try_to_wake_up()`：记录唤醒时刻与唤醒者栈。"""
        self.wakets[tid] = ts
        self.waker_stack[tid] = waker_stack

    def on_switch_finish(
        self, prev_tid: int, prev_state: int, cur_tid: int, cur_stack: str, ts: float
    ) -> Optional[Sample]:
        """`finish_task_switch(prev)` 结尾：给 prev 打起点，结算 cur 的等待。"""
        # ① prev 开始睡
        self.sleeptime[prev_tid] = ts
        self.state[prev_tid] = prev_state
        # ② cur 若是被唤醒后才跑到这里，结算它这一段 off-CPU
        start = self.sleeptime.get(cur_tid)
        if not start:  # 0 或不存在 ⇒ 本次不是阻塞后的唤醒
            return None
        state = self.state.get(cur_tid, TASK_INTERRUPTIBLE)
        sample = Sample(
            tid=cur_tid,
            start=start,
            end=ts,
            state=state,
            stack=cur_stack,
            woken_at=self.wakets.pop(cur_tid, None),
            waker_stack=self.waker_stack.pop(cur_tid, ""),
        )
        self.sleeptime[cur_tid] = 0  # 原文：清零而不是删除
        self.samples.append(sample)
        return sample

    # ---------------------------------------------------------------- 查询 ---
    def filtered(self) -> List[Sample]:
        if self.state_filter is None:
            return list(self.samples)
        return [s for s in self.samples if s.state == self.state_filter]

    def by_stack(self, samples: Optional[Sequence[Sample]] = None) -> Dict[str, float]:
        """按阻塞栈聚合 off-CPU 时间（折叠格式的 `count` 列）。"""
        out: Dict[str, float] = {}
        for s in (self.filtered() if samples is None else samples):
            out[s.stack] = out.get(s.stack, 0.0) + s.off_cpu
        return out

    def by_stack_count(self, samples: Optional[Sequence[Sample]] = None) -> Dict[str, int]:
        out: Dict[str, int] = {}
        for s in (self.filtered() if samples is None else samples):
            out[s.stack] = out.get(s.stack, 0) + 1
        return out

    def request_sync(self, marker: str) -> List[Sample]:
        """线程池陷阱的解药：只保留含请求上下文标记的栈（等价 `grep do_command`）。"""
        return [s for s in self.filtered() if marker in s.stack]

    def off_wake(self, samples: Optional[Sequence[Sample]] = None) -> Dict[Tuple[str, str], float]:
        """off-wake 联合归因：键是 (唤醒者栈, 阻塞栈)，值仍是 off-CPU 时间。"""
        out: Dict[Tuple[str, str], float] = {}
        for s in (self.filtered() if samples is None else samples):
            k = (s.waker_stack or "(unknown waker)", s.stack)
            out[k] = out.get(k, 0.0) + s.off_cpu
        return out


def folded(stack: str, count: float) -> str:
    """折叠格式：一行一个栈，分号分隔，末尾是指标值。"""
    return f"{stack} {count:g}"


def parse_folded(lines: Sequence[str]) -> Dict[str, float]:
    out: Dict[str, float] = {}
    for ln in lines:
        ln = ln.strip()
        if not ln:
            continue
        frames, _, cnt = ln.rpartition(" ")
        out[frames] = out.get(frames, 0.0) + float(cnt)
    return out


def parse_differential(lines: Sequence[str]) -> Dict[str, Tuple[float, float]]:
    """difffolded.pl 的三列输出：`stack before after`（本 demo 只用它做回归比对）。"""
    out: Dict[str, Tuple[float, float]] = {}
    for ln in lines:
        parts = ln.rstrip().rsplit(" ", 2)
        if len(parts) != 3:
            continue
        out[parts[0]] = (float(parts[1]), float(parts[2]))
    return out


def elapsed_split(sample: Sample) -> Dict[str, float]:
    """把一个 off-CPU 区间拆成「真阻塞 / 调度延迟 / 膨胀比」三段。"""
    return {
        "off_cpu": sample.off_cpu,
        "blocked": sample.blocked,
        "sched_latency": sample.sched_latency,
        "inflation": sample.inflation,
    }
